File: mahjong_utils/models/hand_pattern.py
from abc import abstractmethod, ABC
from typing import Iterable, Optional, Tuple, FrozenSet

from pydantic import Field
from pydantic.main import BaseModel

from mahjong_utils.models.furo import Furo, Kan
from mahjong_utils.models.mentsu import Mentsu, Shuntsu, Kotsu
from mahjong_utils.models.tatsu import Tatsu
from mahjong_utils.models.tile import Tile


class HandPattern(BaseModel, ABC):
    @property
    @abstractmethod
    def menzen(self) -> bool:
        raise NotImplementedError()

    @property
    @abstractmethod
    def tiles(self) -> Iterable[Tile]:
        raise NotImplementedError()

    def __hash__(self):
        return hash(self.__class__) + hash(tuple(self.__dict__.values()))

    @abstractmethod
    def __encode__(self) -> dict:
        raise NotImplementedError()

    @classmethod
    def __decode__(cls, data: dict) -> "HandPattern":
        if data['type'] == 'RegularHandPattern':
            return RegularHandPattern.__decode__(data)
        elif data['type'] == 'ChitoiHandPattern':
            return ChitoiHandPattern.__decode__(data)
        elif data['type'] == 'KokushiHandPattern':
            return KokushiHandPattern.__decode__(data)
        else:
            raise ValueError("invalid type: " + data['type'])


class RegularHandPattern(HandPattern):
    """
    表示一个结构分析后的以标准形为目标的手牌
    """

    k: int
    jyantou: Optional[Tile] = None
    menzen_mentsu: Tuple[Mentsu, ...] = Field(default_factory=tuple)
    furo: Tuple[Furo, ...] = Field(default_factory=tuple)
    tatsu: Tuple[Tatsu, ...] = Field(default_factory=tuple)
    remaining: Tuple[Tile, ...] = Field(default_factory=tuple)

    def __encode__(self) -> dict:
        return dict(
            type="RegularHandPattern",
            k=self.k,
            jyantou=self.jyantou.__encode__() if self.jyantou is not None else None,
            menzenMentsu=[mt.__encode__() for mt in self.menzen_mentsu],
            furo=[fr.__encode__() for fr in self.furo],
            tatsu=[tt.__encode__() for tt in self.tatsu],
            remaining=[t.__encode__() for t in self.remaining],
        )

    @classmethod
    def __decode__(cls, data: dict) -> "RegularHandPattern":
        return RegularHandPattern(
            k=data["k"],
            jyantou=Tile.__decode__(data["jyantou"]) if data["jyantou"] is not None else None,
            menzen_mentsu=tuple(Mentsu.__decode__(x) for x in data["menzenMentsu"]),
            furo=tuple(Furo.__decode__(x) for x in data["furo"]),
            tatsu=tuple(Tatsu.__decode__(x) for x in data["tatsu"]),
            remaining=tuple(Tile.__decode__(x) for x in data["remaining"])
        )

    @property
    def mentsu(self) -> Iterable[Mentsu]:
        for mt in self.menzen_mentsu:
            yield mt
        for fr in self.furo:
            yield fr

    @property
    def shuntsu(self) -> Iterable[Shuntsu]:
        for mt in self.mentsu:
            if isinstance(mt, Shuntsu):
                yield mt

    @property
    def kotsu(self) -> Iterable[Kotsu]:
        for mt in self.mentsu:
            if isinstance(mt, Kotsu):
                yield mt

    @property
    def anko(self) -> Iterable[Kotsu]:
        for mt in self.menzen_mentsu:
            if isinstance(mt, Kotsu):
                yield mt
        for fr in self.furo:
            if isinstance(fr, Kan) and fr.ankan:
                yield fr

    @property
    def menzen(self) -> bool:
        for fr in self.furo:
            if not isinstance(fr, Kan) or not fr.ankan:
                return False
        return True

    @property
    def tiles(self) -> Iterable[Tile]:
        if self.jyantou is not None:
            yield self.jyantou
            yield self.jyantou
        for mt in self.mentsu:
            for t in mt.tiles:
                yield t
        for tt in self.tatsu:
            yield tt.first
            yield tt.second
        for t in self.remaining:
            yield t


class ChitoiHandPattern(HandPattern):
    """
    表示一个结构分析后的以七对子为目标的手牌
    """

    pairs: FrozenSet[Tile] = Field(default_factory=frozenset)
    remaining: Tuple[Tile, ...] = Field(default_factory=tuple)

    def __encode__(self) -> dict:
        return dict(
            type="ChitoiHandPattern",
            pairs=[t.__encode__() for t in self.pairs],
            remaining=[t.__encode__() for t in self.remaining],
        )

    @classmethod
    def __decode__(cls, data: dict) -> "ChitoiHandPattern":
        return ChitoiHandPattern(
            pairs=frozenset(Tile.__decode__(x) for x in data["pairs"]),
            remaining=tuple(Tile.__decode__(x) for x in data["remaining"])
        )

    @property
    def menzen(self) -> bool:
        return True

    @property
    def tiles(self) -> Iterable[Tile]:
        for p in self.pairs:
            yield p
            yield p
        for t in self.remaining:
            yield t


class KokushiHandPattern(HandPattern):
    """
    表示一个结构分析后的以国士无双为目标的手牌
    """

    yaochu: FrozenSet[Tile] = Field(default_factory=frozenset)
    repeated: Optional[Tile] = None
    remaining: Tuple[Tile, ...] = Field(default_factory=tuple)

    def __encode__(self) -> dict:
        return dict(
            type="KokushiHandPattern",
            yaochu=[t.__encode__() for t in self.yaochu],
            repeated=self.repeated.__encode__() if self.repeated is not None else None,
            remaining=[t.__encode__() for t in self.remaining],
        )

    @classmethod
    def __decode__(cls, data: dict) -> "KokushiHandPattern":
        return KokushiHandPattern(
            yaochu=frozenset(Tile.__decode__(x) for x in data["yaochu"]),
            repeated=Tile.__decode__(data["repeated"]) if data["repeated"] is not None else None,
            remaining=tuple(Tile.__decode__(x) for x in data["remaining"])
        )

    @property
    def menzen(self) -> bool:
        return True

    @property
    def tiles(self) -> Iterable[Tile]:
        for t in self.yaochu:
            if self.repeated == t:
                yield t
            yield t
        for t in self.remaining:
            yield t

    __all__ = ("HandPattern", "RegularHandPattern", "ChitoiHandPattern", "KokushiHandPattern")

File: src/commonMain/kotlin/mahjongutils/models/hand/HandPattern.kt
package mahjongutils.models.hand

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import mahjongutils.models.*

/**
 * 手牌形
 */
interface HandPattern : IHasFuro {
    /**
     * 手牌（包括门前与副露）
     */
    val tiles: List<Tile>

    /**
     * 浮牌
     */
    val remaining: List<Tile>
}

// 单纯是为了sealed
@Serializable
sealed interface CommonHandPattern : HandPattern

/**
 * 以标准形为目标的手牌
 */
interface IRegularHandPattern : HandPattern {
    /**
     * 目标面子组数（=手牌数/4）
     */
    val k: Int

    /**
     * 雀头
     */
    val jyantou: Tile?

    /**
     * 门前面子
     */
    val menzenMentsu: List<Mentsu>

    /**
     * 搭子
     */
    val tatsu: List<Tatsu>

    override val tiles: List<Tile>
        get() = buildList {
            jyantou?.let {
                add(it)
                add(it)
            }
            for (mt in menzenMentsu) {
                addAll(mt.tiles)
            }
            for (fr in furo) {
                addAll(fr.asMentsu().tiles)
            }
            for (tt in tatsu) {
                add(tt.first)
                add(tt.second)
            }
            addAll(remaining)
        }

    /**
     * 面子（包括门前与副露）
     */
    val mentsu: List<Mentsu>
        get() = menzenMentsu + furo.map { it.asMentsu() }

    /**
     * 暗刻
     */
    val anko: List<Kotsu>
        get() = menzenMentsu.filterIsInstance<Kotsu>() +
                furo.filterIsInstance<Kan>().filter { it.ankan }.map { it.asMentsu() }
}

/**
 * 以标准形为目标的手牌
 */
@Serializable
@SerialName("RegularHandPattern")
data class RegularHandPattern(
    override val k: Int,
    override val jyantou: Tile?,
    override val menzenMentsu: List<Mentsu>,
    override val furo: List<Furo>,
    override val tatsu: List<Tatsu>,
    override val remaining: List<Tile>,
) : IRegularHandPattern, CommonHandPattern

/**
 * 以七对子为目标的手牌
 */
interface IChitoiHandPattern : HandPattern {
    /**
     * 已有对子
     */
    val pairs: Set<Tile>

    override val furo: List<Furo>
        get() = emptyList()

    override val tiles: List<Tile>
        get() = buildList {
            addAll(pairs)
            addAll(pairs)
            addAll(remaining)
        }
}

/**
 * 以七对子为目标的手牌
 */
@Serializable
@SerialName("ChitoiHandPattern")
data class ChitoiHandPattern(
    override val pairs: Set<Tile>,
    override val remaining: List<Tile>
) : IChitoiHandPattern, CommonHandPattern

/**
 * 以国士无双为目标的手牌
 */
interface IKokushiHandPattern : HandPattern {
    /**
     * 幺九牌
     */
    val yaochu: Set<Tile>

    /**
     * 重复的幺九牌
     */
    val repeated: Tile?

    override val furo: List<Furo>
        get() = emptyList()

    override val tiles: List<Tile>
        get() = buildList {
            addAll(yaochu)
            repeated?.let { add(it) }
            addAll(remaining)
        }
}

/**
 * 以国士无双为目标的手牌
 */
@Serializable
@SerialName("KokushiHandPattern")
data class KokushiHandPattern(
    override val yaochu: Set<Tile>,
    override val repeated: Tile?,
    override val remaining: List<Tile>
) : IKokushiHandPattern, CommonHandPattern